### DISCUSSION TITLE
80 Columns mode on appleII+

### DISCUSSION THREAD

**Discussion by abaffa** [markadev/AppleII-VGA#27] [Ideas]
Hi there, 
 I'm wondering if we could emulate 80 columns directly to vga using an appleII+. The 80 columns mode is available on apple//e but not in appleII+. Is that possible to emulate an 80 columns card like videx and have the 80 columns mode on an appleII+? What are the differencees between 80 columns mode on apple//e and appleii+?

**Comment by markadev** [comment]
Hey there!  I'm not entirely sure what the difference is with IIe 80 column mode but at a hardware level it's going to be different just due to the way that the IIe 80 column card and AUX memory works which is very specific to the IIe.  I suppose that software would usually just `PR#3` and then print control characters to control the terminal.

I think it is theoretically possible to emulate something like a Videx card for 80 column support but I am not sure if the current hardware revision can support it.  It's something I initially had wanted to do but ultimately abandoned because I didn't find any software that uses it so thought it wasn't so useful.  Of course, Videx cards are selling on eBay so I guess _something_ uses 80 column mode 😀 

The challenge for me in implementing it was that the Pico has to emulate a full device with ROMs and registers on the Apple bus instead of just being a passive listener for bus activity and therefore has to be able to respond to memory read cycles from the 6502 very fast.  I had prototyped this "read cycle" support but it was not reliable and flexible enough to work for something as complex as the Videx.  Ultimately I removed that prototype code and repurposed the GPIO pin but the original code is [still in the git repository](https://github.com/markadev/AppleII-VGA/blob/read-cycle-last-call/pico/abus.pio).  It's possible that it could be made to work after more thought and development iterations.  Other people are also using Picos in this manner in other projects, which is a good sign that it could work.

I do think this would be a great addition though, and is _probably_ doable.

**Reply by abaffa** [reply]
Thanks for your answer. The 80 columns mode was useful in (at least) two situations: BBSes and CP/M programs (visicalc, wordstart, etc). [Videx has a demo disk](https://mirrors.apple2.org.za/ftp.apple.asimov.net/images/hardware/video/Videx%20VideoTerm%20Util%20Disk.dsk) that could be useful during the tests and it demonstrate all video modes and resources. 

I think Apple//e 80columns was based on videx to keep legacy programs working but it runs in a different memory address so I'm not totally sure about its compatibility. And also, It sounds exacly as you said: PR#3 redirects the terminal output to the board and them it starts showing the video output. The videx works in paralel to the original video circuit (exacly as AppleIIVGA) and works only when pr#3 is running (or the system sends data to the memory address that represents it).

I was wondering if we should attach the AppleIIVGA card to slot 3 in order to be compatible to videx and maybe use part of the solution written to apple//e to generate the 80col output.

--

Last year, I wrote a simple [appleII emulator](https://github.com/abaffa/appleII_emulator) to understand how everything works and I could understand how video is processed. I'd like to help on this task but, unfourtunatelly, I need to have some time to understand the project source code. 

By the way, just to justify why I'm insisting on this topic instead of using an apple//e... it's because I designed a [homebrew appleii+ clone](https://hackaday.io/project/193291-apple-homebrew-clone-for-baffa-2-project) following the original schematics and your card was essential to simplify the video circuit. As the AppleII+ is built using TLL, it was a simpler option than trying to understand how to reproduce the "IOU" and "MMU" ICs.- using your card I could simply ignore all video schematics and focus on the address decoder, RAM/ROM and I/O circuits.

**Reply by markadev** [reply]
Ah I see, and that sounds like a cool project.  One of my favorite things about the II+ is that it was built before logic started getting condensed into proprietary ICs so everything is accessible and replaceable.

There was another comment here not too long ago (https://github.com/markadev/AppleII-VGA/discussions/15#discussioncomment-7960575) about II+ 80 column support where we talked about "full Videx emulation" vs just "Videx compatibility" and the latter might be good enough for most people's use.  It would still require having a videx card for the ROM & extra screen RAM but they can still be found on eBay and I've also seen [modern open source clones](https://github.com/btb/80ColumnCard).  I'm not sure if @aaronsmithdev has made any progress yet but this would be the approach I'd recommend for now.

**Reply by aaronsmithdev** [reply]
Sorry @markadev, @abaffa - no progress here and no alternative solution or input. My position was eliminated at work and I've had to hussle with interviews and contract work. I'll be out for the count for at least the next month or so.

Great work Augusto! Thanks for charging forward with this! You've made some ][+ folks very happy!

**Reply by abaffa** [reply]
I'm sorry to hear that.  I hope you find a new position soon.

I'm very happy to have helped with the videx card emulation and I hope this can make other appleii+ users as happy as I am now.

**Comment by abaffa** [comment]
finishing this thread - all done. The code is available at https://github.com/abaffa/AppleII-VGA/ and I requested a pull https://github.com/markadev/AppleII-VGA/pull/38 so Mark can verify my solution and suggest some project standards.

**80 Columns mode on AppleII+ - Videos**

[![AppleII Videx Videoterm 80 Columns card Demo running at AppleIIVga by Markadev](http://img.youtube.com/vi/jKAShbG-p9Y/0.jpg)](https://www.youtube.com/watch?v=jKAShbG-p9Y "Apple][+ Videx Videoterm 80 Columns card Demo running at AppleIIVga by Markadev")

[![AppleII Accessing BBS using Videx Videoterm 80 Columns card running at AppleIIVga by Markadev](http://img.youtube.com/vi/D4BjUacbDPo/0.jpg)](https://www.youtube.com/watch?v=D4BjUacbDPo "Apple][+ Accessing BBS using Videx Videoterm 80 Columns card running at AppleIIVga by Markade")
